src/GCDHelper.cs

```diff
@@ -0,0 +1,616 @@
+        public bool Queue_Triangle { get; private set; }
+        public bool SlideStart_VerticalBar { get; private set; }
+        public bool SlideEnd_VerticalBar { get; private set; }
+        public bool SlideStart_LeftTri { get; private set; }
+        public bool SlideStart_RightTri { get; private set; }
+        public bool SlideEnd_RightTri { get; private set; }
+        public bool Slide_Background { get; private set; }
+        public float Queue_Lock_Start { get; private set; }
+        public float Slide_Bar_Start { get; private set; }
+        public float Slide_Bar_End { get; private set; }
+
+        private float previousPos = 1f;
+        static readonly float epsilon = 0.02f;
+        
+        private BarDecisionHelper() { }
+        public static BarDecisionHelper Instance {
+            get {
+                instance ??= new BarDecisionHelper();
+                return instance;
+            }
+        }
+        public enum BarState {
+            GCDOnly,
+            ShortCast,
+            LongCast,
+            NonAbilityCast,
+            Mount,
+            Idle
+        }
+        public BarState currentState;
+
+        public void Update(BarInfo bar, Configuration conf, GCDHelper helper) {                
+            if (bar.CurrentPos > (epsilon / bar.TotalBarTime) && bar.CurrentPos < previousPos - epsilon) {
+                // Reset
+                previousPos = 0f;
+                ResetBar(conf);
+
+                // Handle Castbar
+                if(bar.IsCastBar){
+                    Slide_Bar_Start = bar.GCDTime_SlidecastStart;
+                    Slide_Bar_End = conf.SlideCastFullBar ? 1f : bar.GCDTotal_SlidecastEnd;
+                    if (bar.IsNonAbility) {
+                        Queue_Lock_Start = 0f;
+                        Queue_VerticalBar = false;
+                        Queue_Triangle = false;
+                        Slide_Bar_End = 1f;
+                        currentState = helper.queuedAbilityActionType switch
+                        {
+                            ActionType.Mount => BarState.Mount,
+                            _ => BarState.NonAbilityCast,
+                        };
+                    }
+                    else if (bar.IsShortCast) {
+                        Queue_Lock_Start = 0.8f;
+                        if (Math.Abs(Slide_Bar_End - Queue_Lock_Start) < epsilon)
+                            Slide_Bar_End = Queue_Lock_Start;
+                        currentState = BarState.ShortCast;
+                    }
+                    else if (!bar.IsShortCast) {
+                        Queue_Lock_Start = 0.8f * (bar.GCDTotal / bar.CastTotal);
+                        currentState = BarState.LongCast;
+                    }
+                }
+                // Handle GCDBar
+                else if (!bar.IsCastBar && !bar.IsShortCast) {
+                    Queue_Lock_Start = Math.Max((bar.GCDTotal - 0.5f - conf.QueueLockPingOffset) / bar.GCDTotal, 0f);
```

Comment:
There are sort of two different concepts at play here:

- Ability Lockout (this prevents a spell from being queued until ~500ms before the end of the current GCD
- Clip Timing (this represents when you can queue a spell without cutting into the next GCD.

In a perfect world, those would always be in sync; but due to latency, the Clip Time (last possible moment to use an oGCD) is always longer than the Ability Lockout (which doesn't depend on ping at all, and seems to be a client feature).  So basically, the slider lets you decide if the visual queuelock is aligned with the ability lockout, or the last moment to weave.  Depending on what class I'm playing I tend to prefer one or the other (my bard I prefer to see the last moment I can weave, my Red Mage I'd rather know the first moment I can queue).

This seemed like a straightforward way to add a stealth feature without changing too much.  I tried representing both lines split apart like I did with the slidecast window, but that ends up being a bit busy.  So this way, if you know what the difference is, you can have one or the other, but not both.

It's ultimately of limited usefulness to probably most people and I just added it because it was simple to implement.  It can go I don't think it's a deal-breaker.

Reply:
Ohh, now I get it. I was just thinking of Ability Lockout. That makes sense.
I agree with removing it for now and focusing on what we have for now. I can see a world where we add in the future configuration options individual for ability Lockout, clip timing and slidecast, including if they are shown and which visualization we use for it. Something like that where the user opts into it while learning about the meaning of the visualization should make it easier to understand.

Reply:
Yeah, I think it's complicated enough now that some kind of configuration wizard / tutorial is probably in order, with some pictures of the bar's parts to show what each component is.  As it is now, people sort of have to turn on/off settings to see what they do.  I'm going to have a pass through the config and add in some tooltips for everything that I can before we merge to try to make it less confusing.